src/pkcs11/pkcs11-spy.c

```diff
@@ -400,6 +400,114 @@ spy_attribute_list_out(const char *name, CK_ATTRIBUTE_PTR pTemplate,
 	print_attribute_list(spy_output, pTemplate, ulCount);
 }
 
+static void
+spy_dump_mechanism(const char *name, CK_MECHANISM_PTR pMechanism)
+{
+	if (!pMechanism) {
+		fprintf(spy_output, "%s=(null)\n", name);
+		return;
+	}
+
+	fprintf(spy_output, "%s->type=%s\n", name, lookup_enum(MEC_T, pMechanism->mechanism));
+	switch (pMechanism->mechanism) {
+	case CKM_AES_GCM:
+		if (pMechanism->pParameter != NULL) {
+			CK_GCM_PARAMS *param =
+				(CK_GCM_PARAMS *) pMechanism->pParameter;
+			spy_dump_string_in("pIv[ulIvLen]",
+				param->pIv, param->ulIvLen);
+			spy_dump_ulong_in("ulIvBits", param->ulIvBits);
+			spy_dump_string_in("pAAD[ulAADLen]",
+				param->pAAD, param->ulAADLen);
+			fprintf(spy_output, "%s->pParameter->ulTagBits=%lu\n", name, param->ulTagBits);
+		} else {
+			fprintf(spy_output, "Parameters block for %s is empty...\n",
+				lookup_enum(MEC_T, pMechanism->mechanism));
```

Comment:
How do you think should these == NULL blocks be changed to like below:
```c
		if (pMechanism->pParameter == NULL) {
			fprintf(spy_output, "[in] %s->pParameter = NULL\n", name);
			break;
		}
```

Reply:
Yes. The current version is more like a coment, but this should be more like key+values list to be easy to search through.